### PR TITLE
delete_config with drives_conversion

### PR DESCRIPTION
**Pull request by Rachit7194**
Hello @cdearborn ,
This PR contains changes related to upgrade delete_configuration and converting drives from raid to jbod after delete configuration with new testcases.

### CODE REVIEW

File: ironic/drivers/modules/drac/common.py [hunk 1]
```diff
@@ -115,3 +115,8 @@ def get_drac_client(node):
                                     driver_info['drac_protocol'])
 
     return client
+
+
+def get_reboot_options():
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/common.py]: It looks like this change is due to me resyncing my fork with Ironic.  Can you undo it?

File: ironic/drivers/modules/drac/raid.py [hunk 2]
```diff
@@ -293,6 +293,43 @@ def clear_foreign_config(node, raid_controller):
                    'error': exc})
         raise exception.DracOperationError(error=exc)
 
+def change_physical_disk_state(node, mode,
+                               controllers_to_physical_disk_ids=None):
+    """
+    :param node: an ironic node objec
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/raid.py]: Can you change "objec" to "object" in the line above?

File: ironic/drivers/modules/drac/raid.py [hunk 3]
```diff
@@ -293,6 +293,43 @@ def clear_foreign_config(node, raid_controller):
...
+    try:
+
+        drac_job.validate_job_queue(node)
+
+        client = drac_common.get_drac_client(node)
+        return client.change_physical_disk_state(
+            mode,
+            controllers_to_physical_disk_ids=controllers_to_physical_disk_ids)
+    except drac_exceptions.BaseClientException as exc:
+        LOG.error('DRAC driver failed to free foreign driver '
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/raid.py]: This error message references freeing foreign drives.  Looks like it  needs to be updated.

File: ironic/drivers/modules/drac/raid.py [hunk 4]
```diff
@@ -562,8 +599,8 @@ def _find_configuration(logical_disks, physical_disks):
                                         physical_disks_by_type,
                                         free_space_mb))
             if not result:
-                error_msg = _('failed to find matching physical disks for all '
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/raid.py]: Upstream reviewers will probably ask you to revert this change since it's outside the scope of this patch.

File: ironic/drivers/modules/drac/raid.py [hunk 5]
```diff
@@ -348,7 +385,7 @@ def _calculate_spans(raid_level, disks_count):
     """Calculates number of spans for a RAID level given a physical disk count
 
     :param raid_level: RAID level of the virtual disk.
-    :param disk_count: number of physical disks used for the virtual disk.
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/raid.py]: Upstream reviewers will probably ask you to revert this change since it's outside the scope of this patch.

File: ironic/drivers/modules/drac/raid.py [hunk 6]
```diff
@@ -645,18 +682,18 @@ def _assign_disks_to_volume(logical_disks, physical_disks_by_type,
          interface_type, size_mb), disks in physical_disks_by_type.items():
 
         if ('disk_type' in logical_disk
-            and logical_disk['disk_type'] != disk_type):
+                and logical_disk['disk_type'] != disk_type):
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/raid.py]: Upstream reviewers will probably ask you to revert this change since it's outside the scope of this patch.

File: ironic/drivers/modules/drac/raid.py [hunk 7]
```diff
@@ -645,18 +682,18 @@ def _assign_disks_to_volume(logical_disks, physical_disks_by_type,
...
             continue
 
         if ('interface_type' in logical_disk
-            and logical_disk['interface_type'] != interface_type):
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/raid.py]: Upstream reviewers will probably ask you to revert this change since it's outside the scope of this patch.

File: ironic/drivers/modules/drac/raid.py [hunk 8]
```diff
@@ -645,18 +682,18 @@ def _assign_disks_to_volume(logical_disks, physical_disks_by_type,
...
+                and logical_disk['interface_type'] != interface_type):
             continue
 
         # filter out disks without free disk space
         disks = [disk for disk in disks if free_space_mb[disk] > 0]
 
-        # sort disks by free size which is important if we have max disks limit
-        # on a volume
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/raid.py]: Upstream reviewers will probably ask you to revert this change since it's outside the scope of this patch.

File: ironic/drivers/modules/drac/raid.py [hunk 9]
```diff
@@ -677,8 +714,9 @@ def _assign_disks_to_volume(logical_disks, physical_disks_by_type,
 
         for disks_count in range(min_disks, candidate_max_disks + 1):
             if ('number_of_physical_disks' in logical_disk
-                and logical_disk['number_of_physical_disks'] != disks_count):
-                    continue
+                    and logical_disk[
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/raid.py]: Upstream reviewers will probably ask you to revert this change since it's outside the scope of this patch.

File: ironic/drivers/modules/drac/common.py (hunk 1 above)
Comment (Rachit7194) [reply ironic/drivers/modules/drac/common.py]: Done

File: ironic/drivers/modules/drac/raid.py (hunk 2 above)
Comment (Rachit7194) [reply ironic/drivers/modules/drac/raid.py]: Done

File: ironic/drivers/modules/drac/raid.py (hunk 3 above)
Comment (Rachit7194) [reply ironic/drivers/modules/drac/raid.py]: Done

File: ironic/drivers/modules/drac/raid.py (hunk 4 above)
Comment (Rachit7194) [reply ironic/drivers/modules/drac/raid.py]: Reverted back to normal.

File: ironic/drivers/modules/drac/raid.py (hunk 5 above)
Comment (Rachit7194) [reply ironic/drivers/modules/drac/raid.py]: Done

File: ironic/drivers/modules/drac/raid.py (hunk 6 above)
Comment (Rachit7194) [reply ironic/drivers/modules/drac/raid.py]: Done

File: ironic/drivers/modules/drac/raid.py (hunk 7 above)
Comment (Rachit7194) [reply ironic/drivers/modules/drac/raid.py]: Done

File: ironic/drivers/modules/drac/raid.py (hunk 8 above)
Comment (Rachit7194) [reply ironic/drivers/modules/drac/raid.py]: Done

File: ironic/drivers/modules/drac/raid.py (hunk 9 above)
Comment (Rachit7194) [reply ironic/drivers/modules/drac/raid.py]: Done

File: ironic/drivers/modules/drac/common.py (hunk 1 above)
Comment (cdearborn) [reply ironic/drivers/modules/drac/common.py]: The change is still showing up here.  Did you forget to push the update?

File: ironic/drivers/modules/drac/raid.py [hunk 10]
```diff
@@ -655,7 +693,7 @@ def _assign_disks_to_volume(logical_disks, physical_disks_by_type,
         # filter out disks without free disk space
         disks = [disk for disk in disks if free_space_mb[disk] > 0]
 
-        # sort disks by free size which is important if we have max disks limit
+        # sort disks by free size which is important if we  have max disks limit
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/raid.py]: Can you undo the above change since it's not directly related to this patch?

File: ironic/drivers/modules/drac/raid.py [hunk 11]
```diff
@@ -294,6 +294,44 @@ def clear_foreign_config(node, raid_controller):
...
+    """
+    :param node: an ironic node object.
+    :param mode: enumeration that indicates the mode
+                     to change the disks to.
+    :param controllers_to_physical_disk_ids:Dictionary of controllers and
+                                            corresponding disk ids to convert
+                                            to the requested mode
+    :return: a dictionary containing:
+             - conversion_results, a dictionary that maps controller ids
+             to the conversion results for that controller.  The
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/raid.py]: Can you make the formatting for conversion results be the same as it is in the _change_physical_disk_conversion() method below?

File: ironic/drivers/modules/drac/raid.py [hunk 12]
```diff
@@ -727,6 +765,45 @@ def _filter_logical_disks(logical_disks, include_root_volume,
...
+                    - The is_reboot_required key with a RebootRequired
+                      enumerated value indicating whether the server must be
+                      rebooted to complete disk conversion.
+    """
+    change_disk_state = change_physical_disk_state(
+        node,
+        mode,
+        controllers_to_physical_disk_ids=controllers_to_physical_disk_ids)
+
+    LOG.info("Change_disk_state return-- : {}".format(str(change_disk_state)))
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/raid.py]: In the line above, would you mind changing "Change_disk_state" to "change_physical_disk_state"?

Comment (cdearborn) [reply ironic/drivers/modules/drac/raid.py]: Probably could remove the "-- " as well.

Comment (cdearborn) [reply ironic/drivers/modules/drac/raid.py]: And this should probably be logged at the debug level.

File: ironic/drivers/modules/drac/raid.py [hunk 13]
```diff
@@ -727,6 +765,45 @@ def _filter_logical_disks(logical_disks, include_root_volume,
     return filtered_disks
 
 
+def _change_physical_disk_conversion(
+        node, mode=None, controllers_to_physical_disk_ids=None):
+    """
+    :param node: an ironic node objec
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/raid.py]: Can you change "objec" to "object" in the line above?

File: ironic/drivers/modules/drac/raid.py [hunk 14]
```diff
@@ -294,6 +294,44 @@ def clear_foreign_config(node, raid_controller):
         raise exception.DracOperationError(error=exc)
 
 
+def change_physical_disk_state(node, mode,
+                               controllers_to_physical_disk_ids=None):
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/raid.py]: To simplify things a little, let's remove the controllers_to_physical-disk_ids on the above line and in the comment below since we always pass None

File: ironic/drivers/modules/drac/raid.py [hunk 15]
```diff
@@ -294,6 +294,44 @@ def clear_foreign_config(node, raid_controller):
...
+             rebooted to complete disk conversion.
+    """
+    try:
+
+        drac_job.validate_job_queue(node)
+
+        client = drac_common.get_drac_client(node)
+        return client.change_physical_disk_state(
+            mode,
+            controllers_to_physical_disk_ids=controllers_to_physical_disk_ids)
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/raid.py]: You could just remove the controllers_to_physical_disk_ids parameter in the line above since the default in python-dracclient is None.

File: ironic/drivers/modules/drac/raid.py [hunk 16]
```diff
@@ -727,6 +765,45 @@ def _filter_logical_disks(logical_disks, include_root_volume,
     return filtered_disks
 
 
+def _change_physical_disk_conversion(
+        node, mode=None, controllers_to_physical_disk_ids=None):
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/raid.py]: Can you remove the controllers_to_physical_disk_ids parameter in the line above to simplify things?

Comment (cdearborn) [reply ironic/drivers/modules/drac/raid.py]: And remove it from the comment below.

File: ironic/drivers/modules/drac/raid.py [hunk 17]
```diff
@@ -727,6 +765,45 @@ def _filter_logical_disks(logical_disks, include_root_volume,
...
+                      True indicating that a config job must be created to
+                      complete disk conversion.
+                    - The is_reboot_required key with a RebootRequired
+                      enumerated value indicating whether the server must be
+                      rebooted to complete disk conversion.
+    """
+    change_disk_state = change_physical_disk_state(
+        node,
+        mode,
+        controllers_to_physical_disk_ids=controllers_to_physical_disk_ids)
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/raid.py]: If the above changes are made, then you could remove controllers_to_physical_disk_ids in the line above.

File: ironic/tests/unit/drivers/modules/drac/test_raid.py [hunk 18]
```diff
@@ -280,6 +280,53 @@ def test_clear_foreign_config_fail(self, mock_validate_job_queue,
...
+            controllers_to_physical_disk_ids=None)
+
+        mock_client.change_physical_disk_state.assert_called_once_with(
+            "JBOD", controllers_to_physical_disk_ids=None)
+
+    @mock.patch.object(drac_raid, 'change_physical_disk_state', spec_set=True,
+                       autospec=True)
+    @mock.patch.object(drac_raid, 'commit_config', spec_set=True,
+                       autospec=True)
+    def test__change_physical_disk_conversion(self,
```
Comment (cdearborn) [inline ironic/tests/unit/drivers/modules/drac/test_raid.py:311]: The method name above indicates that _change_physical_disk_conversion is to be tested...

File: ironic/tests/unit/drivers/modules/drac/test_raid.py [hunk 19]
```diff
@@ -280,6 +280,53 @@ def test_clear_foreign_config_fail(self, mock_validate_job_queue,
...
+        mock_get_drac_client.return_value = mock_client
+
+        expected_disk_state = mock_change_physical_disk_state.return_value = {
+            'is_reboot_required': True,
+            'conversion_results': {
+            'RAID.Integrated.1-1': {'is_reboot_required': 'optional',
+                                    'commit_required': False,
+                                    'is_commit_required': True}},
+            'commit_required_ids': ['RAID.Integrated.1-1'] }        
+        actual_disk_state = drac_raid.change_physical_disk_state(
```
Comment (cdearborn) [inline ironic/tests/unit/drivers/modules/drac/test_raid.py]: The line above really isn't executing anything because drac_raid.change_physical_disk_state() is mocked above.  I think we should be calling _change_physical_disk_conversion() instead

File: ironic/tests/unit/drivers/modules/drac/test_raid.py [hunk 20]
```diff
@@ -280,6 +280,53 @@ def test_clear_foreign_config_fail(self, mock_validate_job_queue,
...
+            self.node,
+            mode="JBOD",
+            controllers_to_physical_disk_ids=None)
+
+        drac_raid.commit_config(self.node, 'RAID.Integrated.1-1', reboot=True,
```
Comment (cdearborn) [inline ironic/tests/unit/drivers/modules/drac/test_raid.py]: This call isn't doing anything since drac_raid.commit_config() is mocked above.  I think this call should be removed.

File: ironic/tests/unit/drivers/modules/drac/test_raid.py [hunk 21]
```diff
@@ -671,7 +718,8 @@ def _generate_physical_disks(self):
                        autospec=True)
     def test_create_configuration(
             self, mock_commit_config, mock_validate_job_queue,
-            mock_list_physical_disks, mock_get_drac_client):
```
Comment (cdearborn) [inline ironic/tests/unit/drivers/modules/drac/test_raid.py]: Can you undo this change since it's not directly related to this patch?

File: ironic/tests/unit/drivers/modules/drac/test_raid.py [hunk 22]
```diff
@@ -1565,16 +1619,26 @@ def test__execute_cleaning_foreign_drives(self,
             'is_reboot_required': constants.RebootRequired.false,
             'is_commit_required': False
         }
+        mock_commit_config.side_effect = ['42']
+        mock_change_physical_disk_state.return_value = {
+            'is_reboot_required': True,
```
Comment (cdearborn) [inline ironic/tests/unit/drivers/modules/drac/test_raid.py:1633]: Since the above key is deprecated, probably should remove here to ensure we don't have any dependencies on it.

File: ironic/tests/unit/drivers/modules/drac/test_raid.py [hunk 23]
```diff
@@ -280,6 +280,53 @@ def test_clear_foreign_config_fail(self, mock_validate_job_queue,
...
+                                              mock_change_physical_disk_state,
+                                              mock_get_drac_client):
+        mock_client = mock.Mock()
+        mock_get_drac_client.return_value = mock_client
+
+        expected_disk_state = mock_change_physical_disk_state.return_value = {
+            'is_reboot_required': True,
+            'conversion_results': {
+            'RAID.Integrated.1-1': {'is_reboot_required': 'optional',
+                                    'commit_required': False,
```
Comment (cdearborn) [inline ironic/tests/unit/drivers/modules/drac/test_raid.py]: commit_required is no longer returned by python-dracclient, so we should delete the line above.

File: ironic/tests/unit/drivers/modules/drac/test_raid.py [hunk 24]
```diff
@@ -1565,16 +1619,26 @@ def test__execute_cleaning_foreign_drives(self,
             'is_reboot_required': constants.RebootRequired.false,
             'is_commit_required': False
         }
+        mock_commit_config.side_effect = ['42']
+        mock_change_physical_disk_state.return_value = {
+            'is_reboot_required': True,
+            'conversion_results': {
+            'RAID.Integrated.1-1': {'is_reboot_required': 'optional',
+                                    'commit_required': False,
```
Comment (cdearborn) [inline ironic/tests/unit/drivers/modules/drac/test_raid.py]: commit_required is no longer returned by python-dracclient, so we should delete the line above.

File: ironic/tests/unit/drivers/modules/drac/test_raid.py [hunk 25]
```diff
@@ -1565,16 +1619,26 @@ def test__execute_cleaning_foreign_drives(self,
...
+                                    'is_commit_required': True}},
+            'commit_required_ids': ['RAID.Integrated.1-1']}
 
         with task_manager.acquire(self.context, self.node.uuid,
                                   shared=False) as task:
             return_value = task.driver.raid._execute_cleaning_foreign_drives(
                 task, self.node)
-            mock_resume.assert_called_once_with(
-                task, 'cleaning', 'continue_node_clean')
+            #mock_resume.assert_called_once_with(
```
Comment (cdearborn) [inline ironic/tests/unit/drivers/modules/drac/test_raid.py]: Should remove these comments

File: ironic/tests/unit/drivers/modules/drac/test_raid.py [hunk 26]
```diff
@@ -1565,16 +1619,26 @@ def test__execute_cleaning_foreign_drives(self,
...
 
         self.assertIsNone(return_value)
-        self.assertNotIn('raid_config_parameters',
-                         self.node.driver_internal_info)
-        self.assertNotIn('raid_config_substep',
-                         self.node.driver_internal_info)
+        self.assertEqual(['42'],
+            self.node.driver_internal_info['raid_config_job_ids'])
+        #self.assertNotIn('raid_config_parameters',
+        #                 self.node.driver_internal_info)
```
Comment (cdearborn) [inline ironic/tests/unit/drivers/modules/drac/test_raid.py]: Should remove these commented out asserts and add asserts on raid_config-substep and raid_config_parameters.

**Review by cdearborn — CHANGES_REQUESTED** [review]
Hey Rachit, this looks pretty good overall.  Had just a few nits, and a recommendation to remove controllers_to_physical_disk_ids.  Looks like the unit tests could still use some work.

File: ironic/drivers/modules/drac/raid.py (hunk 10 above)
Comment (Rachit7194) [reply ironic/drivers/modules/drac/raid.py]: Thanks. removed additional single space.

File: ironic/drivers/modules/drac/raid.py (hunk 11 above)
Comment (Rachit7194) [reply ironic/drivers/modules/drac/raid.py]: Done

File: ironic/drivers/modules/drac/raid.py (hunk 12 above)
Comment (Rachit7194) [reply ironic/drivers/modules/drac/raid.py]: Yes i have added this just for the debug level personally. I will remove this when i raise the actual patch against upstream ironic.

File: ironic/drivers/modules/drac/raid.py (hunk 13 above)
Comment (Rachit7194) [reply ironic/drivers/modules/drac/raid.py]: Done.

File: ironic/drivers/modules/drac/raid.py (hunk 14 above)
Comment (Rachit7194) [reply ironic/drivers/modules/drac/raid.py]: Removed.

File: ironic/drivers/modules/drac/raid.py (hunk 15 above)
Comment (Rachit7194) [reply ironic/drivers/modules/drac/raid.py]: Yes, Removed.

File: ironic/drivers/modules/drac/raid.py (hunk 16 above)
Comment (Rachit7194) [reply ironic/drivers/modules/drac/raid.py]: Done.

File: ironic/drivers/modules/drac/raid.py (hunk 17 above)
Comment (Rachit7194) [reply ironic/drivers/modules/drac/raid.py]: Done

File: ironic/tests/unit/drivers/modules/drac/test_raid.py (hunk 18 above)
Comment (Rachit7194) [reply ironic/tests/unit/drivers/modules/drac/test_raid.py:311]: Done.

File: ironic/tests/unit/drivers/modules/drac/test_raid.py (hunk 23 above)
Comment (Rachit7194) [reply ironic/tests/unit/drivers/modules/drac/test_raid.py]: Done

File: ironic/tests/unit/drivers/modules/drac/test_raid.py (hunk 19 above)
Comment (Rachit7194) [reply ironic/tests/unit/drivers/modules/drac/test_raid.py]: Yes, removed and replaced with actual method `_change_physical_disk_conversion()`

File: ironic/tests/unit/drivers/modules/drac/test_raid.py (hunk 20 above)
Comment (Rachit7194) [reply ironic/tests/unit/drivers/modules/drac/test_raid.py]: Removed.

File: ironic/tests/unit/drivers/modules/drac/test_raid.py (hunk 21 above)
Comment (Rachit7194) [reply ironic/tests/unit/drivers/modules/drac/test_raid.py]: Replaced as is.

File: ironic/tests/unit/drivers/modules/drac/test_raid.py (hunk 22 above)
Comment (Rachit7194) [reply ironic/tests/unit/drivers/modules/drac/test_raid.py:1633]: Removed.

File: ironic/tests/unit/drivers/modules/drac/test_raid.py (hunk 24 above)
Comment (Rachit7194) [reply ironic/tests/unit/drivers/modules/drac/test_raid.py]: Done

File: ironic/tests/unit/drivers/modules/drac/test_raid.py (hunk 25 above)
Comment (Rachit7194) [reply ironic/tests/unit/drivers/modules/drac/test_raid.py]: Removed.

File: ironic/tests/unit/drivers/modules/drac/test_raid.py (hunk 26 above)
Comment (Rachit7194) [reply ironic/tests/unit/drivers/modules/drac/test_raid.py]: removed commented asserts and added asserts required.

File: ironic/drivers/modules/drac/common.py (hunk 1 above)
Comment (cdearborn) [reply ironic/drivers/modules/drac/common.py]: Rachit, can you please undo this change?

Comment (Rachit7194) [reply ironic/drivers/modules/drac/common.py]: Sorry i missed it..removed in latest commit

File: ironic/tests/unit/drivers/modules/drac/test_raid.py [hunk 27]
```diff
@@ -280,6 +280,52 @@ def test_clear_foreign_config_fail(self, mock_validate_job_queue,
...
+                                        mock_get_drac_client):
+        mock_client = mock.Mock()
+        mock_get_drac_client.return_value = mock_client
+        drac_raid.change_physical_disk_state(
+            self.node,
+            mode="JBOD")
+
+        mock_client.change_physical_disk_state.assert_called_once_with(
+            "JBOD")
+
```
Comment (cdearborn) [inline ironic/tests/unit/drivers/modules/drac/test_raid.py:306]: Probably want to also validate that validate_job_queue was called once.
It would be nice to mock up a return value from mock_client.change_physical_disk_state().  You could then take the return value from drac_raid.change_physical_disk_state() and compare it against the expected return value here.

File: ironic/tests/unit/drivers/modules/drac/test_raid.py [hunk 28]
```diff
@@ -280,6 +280,52 @@ def test_clear_foreign_config_fail(self, mock_validate_job_queue,
...
+        mock_get_drac_client.return_value = mock_client
+        mock_commit_config.return_value = '42'
+        mock_change_physical_disk_state.return_value = {
+            'is_reboot_required': True,
+            'conversion_results': {
+                'RAID.Integrated.1-1': {'is_reboot_required': 'optional',
+                                        'is_commit_required': True}},
+                'commit_required_ids': ['RAID.Integrated.1-1']}
+
+        drac_raid._change_physical_disk_conversion(
```
Comment (cdearborn) [inline ironic/tests/unit/drivers/modules/drac/test_raid.py]: Probably should check the return value and assert that it is the right value.

**Review by cdearborn — CHANGES_REQUESTED** [review]
Looking good.  There are a few changes we probably should make in the test cases, then I think this is good to go.

File: ironic/tests/unit/drivers/modules/drac/test_raid.py (hunk 27 above)
Comment (Rachit7194) [reply ironic/tests/unit/drivers/modules/drac/test_raid.py:306]: Yes done. 
1. check `validate_job_queue` called once.
2. compare expected and actual of `change_state_result(`drac_raid.change_physical_disk_state()`) `

File: ironic/tests/unit/drivers/modules/drac/test_raid.py (hunk 28 above)
Comment (Rachit7194) [reply ironic/tests/unit/drivers/modules/drac/test_raid.py]: The only value `drac_raid._change_physical_disk_conversion` can return is `'clean wait'` so updated raid.py accordingly(as earlier it was returning None) and assert whether return value of above line `clean_wait` or not.

**Comment by Rachit7194** [comment]
Thanks i will raise a patch today

**Review by cdearborn — APPROVED** [review]
Looks good!

**Comment by cdearborn** [comment]
Since this has been submitted upstream, I'm going ahead and closing it here.
